### PR TITLE
new tenet: unused-private-const-var-same-package

### PR DESCRIPTION
**Pull request by mullikine**
I'm splitting up my previous pull request into 3 tenets, which I will later put into their own bundle.

### CODE REVIEW

File: tenets/codelingo/go/unused-private-const-var-same-package/codelingo.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+tenets:
+  - name: unused-private-const-var-same-package
+    doc: |
+      Find private const or var which are unused in a given package.
+    flows:
+      codelingo/review:
+        comment: |
+          '{{thingname}}' is private yet it is not referenced anywhere else in its package.
```
Comment (BlakeMScurr) [inline tenets/codelingo/go/unused-private-const-var-same-package/codelingo.yaml]: `thingname`?  😆

File: tenets/codelingo/go/unused-private-const-var-same-package/codelingo.yaml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+          It should probably be removed.
+    query: |
+      import codelingo/ast/go
+
+      go.dir:
+        go.file(depth = any):
+          @ review.vars.myfname
```
Comment (BlakeMScurr) [inline tenets/codelingo/go/unused-private-const-var-same-package/codelingo.yaml]: Is `myfname` used anywhere?

File: tenets/codelingo/go/unused-private-const-var-same-package/codelingo.yaml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+                    @ review.vars.thingname
+                    name as privateVarName
+                    private == "true"
+        exclude:
+          go.file(depth = any):
+            go.call_expr(depth = any):
+              go.args:
+                go.ident:
+                  name == privateVarName
+                  private == "true"
```
Comment (BlakeMScurr) [inline tenets/codelingo/go/unused-private-const-var-same-package/codelingo.yaml]: I think you can shorten it down. Does this work?

```
go.dir:
  go.value_spec(depth = any):
    go.names:
      @ review.comment
      go.ident:
        @ review.vars.thingname
        name as privateVarName
        private == "true"
  exclude:
    go.call_expr(depth = any):
      go.args:
        go.ident:
          name == privateVarName
          private == "true"
```

File: tenets/codelingo/go/unused-private-const-var-same-package/codelingo.yaml (hunk 1 above)
Comment (mullikine) [reply tenets/codelingo/go/unused-private-const-var-same-package/codelingo.yaml]: OK I'll change it. haha.

Comment (waigani) [reply tenets/codelingo/go/unused-private-const-var-same-package/codelingo.yaml]: Yeah, you should use a more descriptive var name e.g. {{globalname}}. Also, comment should be "Private global {{globalname}} is never used".

File: tenets/codelingo/go/unused-private-const-var-same-package/codelingo.yaml (hunk 3 above)
Comment (mullikine) [reply tenets/codelingo/go/unused-private-const-var-same-package/codelingo.yaml]: This is geting slightly different behavior. It also appears to have matched one of the structs, which may have been a good thing if I were attempting to combine the 3 tenets, but I'm separating the logic.

File: tenets/codelingo/go/unused-private-const-var-same-package/codelingo.yaml (hunk 2 above)
Comment (mullikine) [reply tenets/codelingo/go/unused-private-const-var-same-package/codelingo.yaml]: Thanks. I will remove this.

File: tenets/codelingo/go/unused-private-const-var-same-package/codelingo.yaml (hunk 3 above)
Comment (BlakeMScurr) [reply tenets/codelingo/go/unused-private-const-var-same-package/codelingo.yaml]: Oh interesting. I guess that the `go.gen_decl` fact is what defines a variable/const definition. Can you at least remove the `go.file` facts without changing the output?

File: tenets/codelingo/go/unused-private-const-var-same-package/codelingo.yaml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+              go.value_spec:
+                go.names:
+                  @ review.comment
+                  go.ident:
+                    @ review.vars.globalname
+                    name as privateVarName
+                    private == "true"
+        exclude:
+          go.file(depth = any):
+            go.call_expr(depth = any):
```
Comment (BlakeMScurr) [inline tenets/codelingo/go/unused-private-const-var-same-package/codelingo.yaml]: Does it matter that the var/const is used in a function call? If you changed the `foo` function to:
```
func foo() {
    a := privateVarA
    fmt.Printf("%s\n", a)
}
```

I suspect your tenet would suddenly think `privateVarA` was unused.

File: tenets/codelingo/go/unused-private-const-var-same-package/codelingo.yaml (hunk 3 above)
Comment (BlakeMScurr) [reply tenets/codelingo/go/unused-private-const-var-same-package/codelingo.yaml]: You can probably also remove the `go.decls` fact.

Comment (BlakeMScurr) [reply tenets/codelingo/go/unused-private-const-var-same-package/codelingo.yaml]: (this isn't blocking btw, it's just nice to get them concise)

Comment (mullikine) [reply tenets/codelingo/go/unused-private-const-var-same-package/codelingo.yaml]: I will give it a go.

Comment (mullikine) [reply tenets/codelingo/go/unused-private-const-var-same-package/codelingo.yaml]: Sweet as. Getting insight as to which facts exactly are unique to a piece of syntax is what I'm hoping to achieve by making my CLQL a little verbose. Please bear with me :)

**Comment by mullikine** [comment]
After simplifying the "exclude:" node to fix the issue of CLQL not matching a reference to an identifier which was not within a function call, there is no longer anything to differentiate it from the match above. I may need some advice on a workaround for this case.

**Comment by mullikine** [comment]
This is waiting on a platform update to make it to release. I'll close this now and reopen when I can.

**Pull request closed without merging**